ee/watchdog/controller_windows.go

```diff
@@ -0,0 +1,344 @@
+		wc.ServiceEnabledChanged(wc.knapsack.LauncherWatchdogEnabled())
+	}
+}
+
+// Run starts a log publication routine. The purpose of this is to
+// pull logs out of the sqlite database and write them to debug.json so we can
+// use all of the existing log publication and cleanup logic while maintaining a single writer
+func (wc *WatchdogController) Run() error {
+	ctx := context.TODO()
+	ticker := time.NewTicker(time.Minute * 5)
+	defer ticker.Stop()
+
+	for {
+		wc.publishLogs(ctx)
+
+		select {
+		case <-ticker.C:
+			continue
+		case <-wc.interrupt:
+			wc.slogger.Log(ctx, slog.LevelDebug,
+				"interrupt received, exiting execute loop",
+			)
+			return nil
+		}
+	}
+}
+
+func (wc *WatchdogController) publishLogs(ctx context.Context) {
+	// note that there is a small window here where there could be pending logs before watchdog is disabled -
+	// there is no harm in leaving them and we could recover these with the original timestamps if we ever needed.
+	// to avoid endlessly re-processing empty logs while we are disabled, we accept this possibility and exit early here
+	if !wc.knapsack.LauncherWatchdogEnabled() {
+		return
+	}
+
+	logsToDelete := make([]any, 0)
+
+	if err := wc.logPublisher.ForEach(func(rowid, timestamp int64, v []byte) error {
+		logRecord := make(map[string]any)
+		logsToDelete = append(logsToDelete, rowid)
+
+		if err := json.Unmarshal(v, &logRecord); err != nil {
+			wc.slogger.Log(ctx, slog.LevelError,
+				"failed to unmarshal sqlite log",
+				"log", string(v),
+				"err", err,
+			)
+
+			// log the issue but don't return an error, we want to keep processing whatever we can
+			return nil
+		}
+
+		logArgs := make([]slog.Attr, len(logRecord))
+		for k, v := range logRecord {
+			logArgs = append(logArgs, slog.Any(k, v))
+		}
+
+		// re-issue the log, this time with the debug.json writer
+		// pulling out the existing log and re-adding all attributes like this will overwrite
+		// the automatic timestamp creation, as well as the msg and level set below
+		wc.slogger.LogAttrs(ctx, slog.LevelInfo, "", logArgs...)
+
+		return nil
+	}); err != nil {
+		wc.slogger.Log(ctx, slog.LevelError, "iterating sqlite logs", "err", err)
+		return
+	}
+
+	if len(logsToDelete) == 0 { // nothing else to do
+		return
+	}
+
+	wc.slogger.Log(ctx, slog.LevelDebug, "collected logs for deletion", "rowids", logsToDelete)
+
+	if err := wc.logPublisher.DeleteRows(logsToDelete...); err != nil {
+		wc.slogger.Log(ctx, slog.LevelError, "cleaning up published sqlite logs", "err", err)
+	}
+}
+
+func (wc *WatchdogController) Interrupt(_ error) {
+	// Only perform shutdown tasks on first call to interrupt -- no need to repeat on potential extra calls.
+	if wc.interrupted {
+		return
+	}
+
+	wc.logPublisher.Close()
+	wc.interrupted = true
+	wc.interrupt <- struct{}{}
+}
+
+func (wc *WatchdogController) ServiceEnabledChanged(enabled bool) {
+	ctx := context.TODO()
+
+	serviceManager, err := mgr.Connect()
```

Comment:
Are there any operations in this function where we want to throw a `backoff.WaitFor` around them?

I get that if we run into e.g. `serviceDoesNotExistError` then there's no point in retrying -- but wondering if for errors like connecting to the service manager here, a wait+retry could be helpful to get the service into the right state? (I see the call to restart the service already has a backoff around it.)

Reply:
added for the connect call here and the service removal portion later